### PR TITLE
Add AppCDS documentation

### LINKED ISSUE #30821: Packaging type -Dquarkus.package.create-appcds=true isn't documented
### Describe the bug

I've just heard that there is application class data sharing support in Quarkus via `-Dquarkus.package.create-appcds=true`, but it isn't documented. So without anyone explicitly asking/knowing about it there is no way to discover that feature.

### Expected behavior

Find a guide explaining what `-Dquarkus.package.create-appcds=true` does and why I should be using it.

### Actual behavior

No way to find documentation other than looking through the code

```
$ grep -rn SharedArchiveFile
core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:108:                            "run the application jar from its directory and also add the '-XX:SharedArchiveFile=app-cds.jsa' " +
core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:239:        // We use the relative paths because at runtime 'java -XX:SharedArchiveFile=... -jar ...' expects the AppCDS and jar files
core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:242:        // The end result is that users can simply use 'java -XX:SharedArchiveFile=app-cds.jsa -jar app.jar'
core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:243:        javaArgs.add("-XX:SharedArchiveFile=" + appCDSPath.getFileName().toString());
extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java:621:                if (effectiveJvmArgument.startsWith("-XX:SharedArchiveFile")) {
extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java:627:                effectiveJvmArguments.add("-XX:SharedArchiveFile=" + appCDSResult.get().getAppCDS().getFileName().toString());
integration-tests/maven/src/test/java/io/quarkus/maven/it/JarRunnerIT.java:497:                List.of("-XX:SharedArchiveFile=app-cds.jsa", "-Xshare:on", "-Xlog:class+path=info"))
```

Or find the property once you know it exists:
https://quarkus.io/guides/all-config#quarkus-core_quarkus.package.appcds-use-container

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by geoand**
The AppCDS generation feature has been part of Quarkus for many releases (and has been made more effective and more usable through various revisions), but until now was not documented.

Closes: #30821

### CODE REVIEW

File: docs/src/main/asciidoc/appcds.adoc [hunk 1]
```diff
@@ -0,0 +1,95 @@
+////
+This guide is maintained in the main Quarkus repository
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
+////
+= Command Mode Applications
```
Comment (gsmet) [inline docs/src/main/asciidoc/appcds.adoc:6]: The title looks off?

File: docs/src/main/asciidoc/appcds.adoc [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+include::_attributes.adoc[]
+:categories: core, cloud
+:summary: This reference guide explains how to enable AppCDS with Quarkus.
+
+This reference guide explains how to enable Application Class Data Sharing in your Quarkus applications.
+
+== What is Application Class Data Sharing (AppCDS)?
+
+link:https://docs.oracle.com/en/java/javase/11/vm/class-data-sharing.html[Application Class Data Sharing] is a JVM feature that helps reduce the startup time and memory footprint of a JVM application.
```
Comment (gsmet) [inline docs/src/main/asciidoc/appcds.adoc:15]: Given 17 is our default JDK, I would rather use this version of the doc:
```suggestion
link:https://docs.oracle.com/en/java/javase/17/vm/class-data-sharing.html[Application Class Data Sharing] is a JVM feature that helps reduce the startup time and memory footprint of a JVM application.
```

File: docs/src/main/asciidoc/appcds.adoc [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+====
+
+[NOTE]
+====
+Given what was mentioned above about how the application needs to be launched in order for the archive to be built, the question arises of how Quarkus deals with this situation.
+
+The answer is that at application build time, right after the application archive is built, Quarkus launches the application, but only the parts of the launch process that are safe are run.
+More specifically, the application is run up until the steps that actually open sockets or run application logic.
+
+This results in the archive process that one hand is completely safe to generate, but on the other hand not being able to archive every single class that the application might need during boot
```
Comment (gsmet) [inline docs/src/main/asciidoc/appcds.adoc:86]:
```suggestion
This results in the archive process that on one hand is completely safe to generate, but on the other hand not being able to archive every single class that the application might need during boot
```

File: docs/src/main/asciidoc/appcds.adoc [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+(which is why users will get a slightly more effective archive if they manually go through the hoops of generating the AppCDS archive).
+====
+
+=== Usage in containers
+
+When building container images using the `quarkus-container-image-jib` extension, Quarkus automatically takes care of all the steps needed to make generate the archive
```
Comment (gsmet) [inline docs/src/main/asciidoc/appcds.adoc:92]:
```suggestion
When building container images using the `quarkus-container-image-jib` extension, Quarkus automatically takes care of all the steps needed to generate the archive
```

File: docs/src/main/asciidoc/appcds.adoc [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+and make it usable at runtime in the container.
+
+This way, by simply setting `quarkus.package.create-appcds` to `true` the generate container can benefit from a slight reduction in startup time and memory usage.
```
Comment (gsmet) [inline docs/src/main/asciidoc/appcds.adoc:95]:
```suggestion
This way, by simply setting `quarkus.package.create-appcds` to `true` the generated container can benefit from a slight reduction in startup time and memory usage.
```

**Review by gsmet — CHANGES_REQUESTED** [review]
Thanks! I spotted a few minor things.

File: docs/src/main/asciidoc/appcds.adoc (hunk 1 above)
Comment (geoand) [reply docs/src/main/asciidoc/appcds.adoc:6]: OOPS!

File: docs/src/main/asciidoc/appcds.adoc (hunk 2 above)
Comment (geoand) [reply docs/src/main/asciidoc/appcds.adoc:15]: Sure yeah, I was wondering this myself

**Review by gsmet — APPROVED** [review]

File: docs/src/main/asciidoc/appcds.adoc [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+[NOTE]
+====
+Given what was mentioned above about how the application needs to be launched in order for the archive to be built, the question arises of how Quarkus deals with this situation.
+
+The answer is that at application build time, right after the application archive is built, Quarkus launches the application, but only the parts of the launch process that are safe are run.
+More specifically, the application is run up until the steps that actually open sockets or run application logic.
+
+This results in the archive process that on one hand is completely safe to generate, but on the other hand not being able to archive every single class that the application might need during boot
+(which is why users will get a slightly more effective archive if they manually go through the hoops of generating the AppCDS archive).
```
Comment (zakkak) [inline docs/src/main/asciidoc/appcds.adoc:87]: Suggestion:

```suggestion
This results in an archive generation process that on one hand is completely safe, but on the other hand is unable to archive every single class that the application might need at boot time.
As a result, users are expected to get a slightly more effective archive if they manually go through the hoops of generating the AppCDS archive.
```

File: docs/src/main/asciidoc/container-image.adoc [hunk 7]
```diff
@@ -99,6 +99,11 @@ To enable this feature, the property `quarkus.bootstrap.workspace-discovery` nee
 when invoking the build tool, either as a build tool property. Setting this property in `application.properties` will **not** work because
 this property needs to be known very early on in the build process.
 
+==== AppCDS
+
+Quarkus supports generating and including an link:https://docs.oracle.com/en/java/javase/11/vm/class-data-sharing.html[Application Class Data Sharing] archive when generating a container image using Jib.
```
Comment (zakkak) [inline docs/src/main/asciidoc/container-image.adoc:104]: For consistency with appcds.adoc

```suggestion
Quarkus supports generating and including an link:https://docs.oracle.com/en/java/javase/17/vm/class-data-sharing.html[Application Class Data Sharing] archive when generating a container image using Jib.
```

**Review by zakkak — CHANGES_REQUESTED** [review]
Thanks for the guide @geoand 

There are a couple of typos and I made a couple of suggestions which IMO improve the guide. After fixing the typos it's good to go IMO!

File: docs/src/main/asciidoc/appcds.adoc [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+== What is Application Class Data Sharing (AppCDS)?
+
+link:https://docs.oracle.com/en/java/javase/17/vm/class-data-sharing.html[Application Class Data Sharing] is a JVM feature that helps reduce the startup time and memory footprint of a JVM application.
+This is achieved by having the JVM create a pre-processed shared archived of classes that are loaded at startup time. As these classes
+are loaded every time the application starts, AppCDS is a conceptually simple way of improving the application startup time,
+without the application itself having to be coded or configured in a specific way. How much of an improvement depends on many factors, such as the number of classes loaded, the underlying hardware etc.
+
+== Vanilla AppCDS generation
+
+The main downside of using AppCDS is that they are generated by launching the application with special flags and having the JVM generate the archive that can then be used in subsequent application launches.
```
Comment (gastaldi) [inline docs/src/main/asciidoc/appcds.adoc:22]: Why is that a downside? That looks like a upside to me :)

Comment (geoand) [reply docs/src/main/asciidoc/appcds.adoc:22]: If you like making your life harder, sure, it's an upside :)

File: docs/src/main/asciidoc/container-image.adoc (hunk 7 above)
Comment (geoand) [reply docs/src/main/asciidoc/container-image.adoc:104]: Ah, nice catch

File: docs/src/main/asciidoc/appcds.adoc [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+= AppCDS
+include::_attributes.adoc[]
+:categories: core, cloud
+:summary: This reference guide explains how to enable AppCDS with Quarkus.
+
+This reference guide explains how to enable Application Class Data Sharing in your Quarkus applications.
+
+== What is Application Class Data Sharing (AppCDS)?
+
+link:https://docs.oracle.com/en/java/javase/17/vm/class-data-sharing.html[Application Class Data Sharing] is a JVM feature that helps reduce the startup time and memory footprint of a JVM application.
```
Comment (jerboaa) [inline docs/src/main/asciidoc/appcds.adoc:15]: Shouldn't this point to `https://docs.oracle.com/en/java/javase/17/docs/specs/man/java.html#application-class-data-sharing` ? There is CDS (Class Data Sharing) and *Application* Class Data Sharing.

File: docs/src/main/asciidoc/appcds.adoc (hunk 8 above)
Comment (gastaldi) [reply docs/src/main/asciidoc/appcds.adoc:22]: I mean that this statement looks like an upside, not the process ;)

File: docs/src/main/asciidoc/appcds.adoc (hunk 9 above)
Comment (geoand) [reply docs/src/main/asciidoc/appcds.adoc:15]: Sure, why not

Comment (geoand) [reply docs/src/main/asciidoc/appcds.adoc]: Sure, fixed

File: docs/src/main/asciidoc/appcds.adoc (hunk 8 above)
Comment (geoand) [reply docs/src/main/asciidoc/appcds.adoc]: How about now?

**Review by zakkak — APPROVED** [review]

File: docs/src/main/asciidoc/appcds.adoc (hunk 8 above)
Comment (gastaldi) [reply docs/src/main/asciidoc/appcds.adoc]: Looks good. Thanks!

**Review by gastaldi — APPROVED** [review]

**Pull request merged**